chalk-solve/src/infer/unify.rs

```diff
@@ -101,7 +101,10 @@ impl<'t, I: Interner> Unifier<'t, I> {
         match (a.data(interner), b.data(interner)) {
             // Unifying two inference variables: unify them in the underlying
             // ena table.
-            (&TyData::InferenceVar(var1), &TyData::InferenceVar(var2)) => {
+            (
+                &TyData::InferenceVar(var1, kind1),
+                &TyData::InferenceVar(var2, kind2),
+            ) if kind1 == kind2 => {
```

Comment:
Nit: Maybe this would be nicer if we combined the three cases and moved the `if` into the arm?

I'm imagining

```rust
(
                &TyData::InferenceVar(var1, kind1),
                &TyData::InferenceVar(var2, kind2),
            ) => {
    if kind1 == kind2 {
        ...
    } else if kind1.is_general() || kind2.is_general() {
        ...
    } else {
        Err(...)
    }
}
```